First CLB batch of changes

georgeedward2000 opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

This PR adds foundational configuration and constants to support the upcoming CLB DiffTracker Engine and pod IP-based backend pools. It includes:

1. **New Load Balancer SKU constant**: Add `LoadBalancerSKUService` (`"service"`) for the service load balancer SKU.

2. **Fix constant naming**: Rename `LoadBalancerBackendPoolConfigurationTypePODIP` -> `LoadBalancerBackendPoolConfigurationTypePodIP` to follow Go naming conventions, and remove the "TODO: support pod IP in the future" comment as this is now being implemented.

3. **Configuration additions** (`pkg/provider/config/azure.go`):
   - Add `ServiceGatewayEnabled` config field to indicate whether the service gateway is enabled for the cluster.
   - Add helper methods: `IsLBBackendPoolTypePodIP()`, `UseServiceLoadBalancer()`, `IsLBBackendPoolTypePodIPAndUseServiceLoadBalancer()`.

4. **Service Gateway initialization logic** (`pkg/provider/azure.go`):
   - Add validation in `InitializeCloudFromConfig` that enables the service gateway only when all 3 conditions are met: `ServiceGatewayEnabled` is true, LB SKU is `service`, and backend pool type is `podIP`.

5. **Cleanup**: Remove unused `TagNameInterconnectGroup` constant.

This PR is the first in a series that will enable pod IP-based backend pools and the CLB DiffTracker Engine. All changes are backward compatible and do not affect existing functionality.

#### Which issue(s) this PR fixes:

Part of #10428.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

georgeedward2000 committed:
First CLB batch of changes

nilo19 commented on pkg/provider/securitygroup/securitygroup.go:
```diff
@@ -304,8 +338,12 @@ func (helper *RuleHelper) RemoveDestinationFromRules(
 	for _, rule := range helper.rules {
-		if !IsManagedSecurityRule(rule) {
-			logger.V(4).Info("Skip rule with not-managed priority", "rule-name", *rule.Name)
+		if rule.Properties.Priority == nil {
```
can you elaborate this change?

georgeedward2000 committed:
Addressing the comments by:
1. Removing NSG code for now
2. Remove unused scaffold code
3. Added context based logger

georgeedward2000 committed:
Addressing the comments by:
1. Removing NSG code for now
2. Remove unused scaffold code
3. Added context based logger

nilo19 commented on pkg/provider/azure.go:
```diff
@@ -261,6 +261,18 @@ func (az *Cloud) InitializeCloudFromConfig(ctx context.Context, config *azurecon
+	// 1. ServiceGatewayEnabled is true
+	// 2. lb sku is service
+	// 3. backendPoolType is PodIP
+	if az.ServiceGatewayEnabled && az.IsLBBackendPoolTypePodIPAndUseServiceLoadBalancer() {
```
Why do we need ServiceGatewayEnabled here? If the user wants to use containerLB, he has set lbsku=service and backendpooltype=podIP. Now he also needs to set ServiceGatewayEnabled=true, which seems a bit complicated to toggle this feature?

david-kow replied:
For AKS side API we only have backendpooltype and lbsku, and the ServiceGatewayEnabled that is passed to Cloud Provider is derived from that.

For Cloud Provider we wanted to have a clear and explicit indication (both on Cloud Provider config and in the code) that the behavior should be shifted to the new code paths, so that just glancing the config tells us whether Service Gateway is disabled on the cluster. I don't feel super strongly about this, so we can get rid of that if you feel it complicates things or is redundant

nilo19 replied:
I'm fine with this, just need to call this out clearly in the future doc pr (from the documentation branch of this repo).

nilo19 commented:
/ok-to-test

nilo19 commented:
/retest
/lgtm
/approve

nilo19 commented:
/lgtm

nilo19 commented:
@georgeedward2000 when the introduced configs are effective, we need to update the release note in that PR. For now, no need to have the release note.
/release-note-none

Liunardy commented:
/cherrypick release-1.36
